### ISSUE TITLE
range.second - range.first == t.size() INTERNAL ASSERT FAILED

### ISSUE THREAD

**Issue by Dungarian** [pytorch/pytorch#38869]
This is my first entry about a problem. So, please feel free to ask anything to clarify the problem. 

## 🐛 Bug
I run the same code in three different machines separately. In two of them, I encounter an error when running "loss.backward()" function. The networks never learn. 

## To Reproduce

1. Run the code for dueling double deep q-learning (Dueling DDQN) and/or DDQNs. 
2. Simulation runs until the number of samples in training sample buffer is enough to train the networks. 
3. Backpropagation function is called and code crashed. 

The error message is as below: 

dueling_ddqn_agent.py in learn(self)
    149 
    150         loss = self.q_eval.loss(q_target, q_pred).to(self.q_eval.device)
--> 151         loss.backward()
    152         self.q_eval.optimizer.step()
    153         self.learn_step_counter += 1

~\Anaconda3\lib\site-packages\torch\tensor.py in backward(self, gradient, retain_graph, create_graph)
    196                 products. Defaults to ``False``.
    197         """
--> 198         torch.autograd.backward(self, gradient, retain_graph, create_graph)
    199 
    200     def register_hook(self, hook):

~\Anaconda3\lib\site-packages\torch\autograd\__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables)
     98     Variable._execution_engine.run_backward(
     99         tensors, grad_tensors, retain_graph, create_graph,
--> 100         allow_unreachable=True)  # allow_unreachable flag
    101 
    102 

RuntimeError: range.second - range.first == t.size() INTERNAL ASSERT FAILED at ..\torch\csrc\autograd\generated\Functions.cpp:57, please report a bug to PyTorch. inconsistent range for TensorList output (copy_range at ..\torch\csrc\autograd\generated\Functions.cpp:57)
(no backtrace available)

## Expected behavior
In one of the computers agents learn as expected. The same code runs without any errors. 

## Environment
PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Microsoft Windows 10 Home Single Language
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 750 Ti
Nvidia driver version: 441.12
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.4
[pip] numpydoc==0.9.1
[pip] torch==1.5.0
[pip] torchvision==0.6.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      245
[conda] mkl-service               2.3.0            py37hb782905_0
[conda] mkl_fft                   1.0.15           py37h14836fe_0
[conda] mkl_random                1.1.0            py37h675688f_0
[conda] numpy                     1.17.4           py37h4320e6b_0
[conda] numpy-base                1.17.4           py37hc3f5095_0
[conda] numpydoc                  0.9.1                      py_0
[conda] torch                     1.5.0                    pypi_0    pypi
[conda] torchvision               0.6.0                    pypi_0    pypi

## Additional context
The same problem occurs with Linear and Conv2d layers. 

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

**Comment by unixpickle** [comment]
I ran into the same internal assertion failure today on `1.5.0`, doing something much simpler with no NNs. The issue does not occur on torch `1.4.0`.

```python
import torch as th

t1 = th.tensor([[1, 2], [3,4]]).float().requires_grad_()
t2 = th.argmin(t1, dim=1)
t3 = t1[t2]
t3.sum().backward()
```

The above produces:

```
RuntimeError: range.second - range.first == t.size() INTERNAL ASSERT FAILED at ../torch/csrc/autograd/generated/Functions.cpp:57, please report a bug to PyTorch. 
```

However, changing the code like so fixes the issue (note the `detach()`):

```python
import torch as th

t1 = th.tensor([[1, 2], [3,4]]).float().requires_grad_()
t2 = th.argmin(t1, dim=1)
t3 = t1[t2.detach()]
t3.sum().backward()
```

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/issues/38982 looks like another occurrence of this problem

**Comment by albanD** [comment]
This looks like a bad case of argmin returning differentiable integers.
This was fixed in https://github.com/pytorch/pytorch/pull/37789 and should be fixed on master. Can you try again with the nightly build and make sure it works fine (I can't repro on master)?

**Comment by gchanan** [comment]
@albanD is going to verify whether this worked in 1.4.  Assuming it did, we'll get something that fixes it into 1.5.1.  Otherwise we'll probably wait until 1.6.0.

**Comment by gchanan** [comment]
Fixed by https://github.com/pytorch/pytorch/pull/39257.
